src/ethereum/frontier/vm/instructions/system.py

```diff
@@ -107,3 +109,154 @@ def return_(evm: Evm) -> None:
     )
     # HALT the execution
     evm.running = False
+
+
+def call(evm: Evm) -> None:
+    """
+    Message-call into an account.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    from ethereum.frontier.vm.interpreter import (
+        STACK_DEPTH_LIMIT,
+        process_message,
+    )
+
+    gas = pop(evm.stack)
+    to = to_address(pop(evm.stack))
+    value = pop(evm.stack)
+    memory_input_start_position = Uint(pop(evm.stack))
+    memory_input_size = pop(evm.stack)
+    memory_output_start_position = Uint(pop(evm.stack))
+    memory_output_size = pop(evm.stack)
+
+    call_gas_fee = calculate_call_gas_cost(evm.env.state, gas, to, value)
+    message_call_gas_fee = calculate_message_call_gas_cost(gas, value)
+
+    evm.gas_left = subtract_gas(evm.gas_left, call_gas_fee)
+
+    gas_input_memory = calculate_gas_extend_memory(
+        evm.memory, memory_input_start_position, memory_input_size
+    )
+    evm.gas_left = subtract_gas(evm.gas_left, gas_input_memory)
+    extend_memory(evm.memory, memory_input_start_position, memory_input_size)
+    gas_output_memory = calculate_gas_extend_memory(
+        evm.memory, memory_output_start_position, memory_output_size
+    )
+    evm.gas_left = subtract_gas(evm.gas_left, gas_output_memory)
+    extend_memory(evm.memory, memory_output_start_position, memory_output_size)
+    call_data = memory_read_bytes(
+        evm.memory, memory_input_start_position, memory_input_size
+    )
+    sender_balance = get_account(
+        evm.env.state, evm.message.current_target
+    ).balance
+    if sender_balance < value:
+        push(evm.stack, U256(0))
+        evm.gas_left += message_call_gas_fee
```

Comment:
Should we have an `add_gas` (or possibly `return_gas`) for symmetry with `subtract_gas`?

Probably not, since returning gas can't raise an exception, but I'm curious what your opinion is.

Reply:
As you pointed out, I would add an `add_gas` method if there was some extra logic to be handled like raising an exception. 
```python
evm.gas_left = add_gas(evm.gas_left, gas_mount)
```
vs 
```python
evm.gas_left += gas_amount
```
I don't see any added benefit between the two examples mentioned above. 


----

I would have added a method if the `gas_left` was mutated inside the method `return_gas`
```python
  evm.return_gas(gas_amount) # this call would automatically increase evm.gas_left by gas_amount
```